### PR TITLE
[#12913] fix(clickhouse): preserve quoted commas in table settings

### LINKED ISSUE #12913: [Bug report] ClickHouse catalog truncates quoted-comma table SETTINGS during load
### Version

main branch

### Describe what's wrong

The ClickHouse JDBC catalog reads table-level settings from `system.tables.engine_full` and exposes them as `settings.*` table properties, but `ClickHouseTableOperations.parseSettingsClause()` currently splits the SETTINGS clause at every comma.

ClickHouse String settings can legally contain commas. For example, `merge_workload = 'gravitino,quoted,comma'` is preserved completely in `system.tables.engine_full`, but Gravitino loads the corresponding property as only `'gravitino`. A later table recreation therefore cannot preserve the original setting value.

The write path already treats each `settings.*` value as one complete ClickHouse scalar literal, so the problem is limited to readback tokenization.

### Error message and/or stacktrace

No exception or stack trace is produced. The metadata value is silently truncated.

```text
Expected: settings.merge_workload = 'gravitino,quoted,comma'
Actual:   settings.merge_workload = 'gravitino
```

### How to reproduce

1. Run Gravitino from the main branch with ClickHouse 24.8.14.
2. Create a native ClickHouse table with a quoted-comma String setting:

```sql
CREATE TABLE test_settings (id Int32)
ENGINE = MergeTree
ORDER BY id
SETTINGS merge_workload = 'gravitino,quoted,comma';
```

3. Query `system.tables.engine_full` and confirm it contains the complete value `'gravitino,quoted,comma'`.
4. Load the table through the Gravitino ClickHouse catalog and inspect its properties.
5. Observe that `settings.merge_workload` is truncated to `'gravitino`.
6. Recreate another table from the loaded setting property and observe that the original scalar literal cannot round-trip correctly.

### Additional context

Issue #11880 and PR #11885 added SETTINGS readback for simple scalar values, but the parser assumes that every comma separates two settings. This issue is a focused parser-correctness follow-up.

The intended fix is limited to splitting on top-level commas and locating the first top-level `=` while ignoring delimiters inside quoted text and nested parentheses. It should preserve the server-returned scalar-literal text, retain existing `settings.*` keys and simple-setting behavior, and fail clearly on structurally malformed metadata without exposing setting values.

The scope does not include ALTER settings from #12768/#12772, column-level SETTINGS, CODEC or TTL parsing, setting-name/type validation, public API changes, or a general ClickHouse SQL parser.

The same truncation was independently reproduced on ClickHouse 25.3.2.39 with `columns_to_prewarm_mark_cache = 'id,name,val'`, confirming that the problem is not specific to the 24.8 project fixture.

Regression coverage should include direct parsing tests for quoted commas, escaping, doubled quotes, nested parentheses, malformed fragments, and duplicate keys, plus one real ClickHouse native-load and load/recreate test.

### PR DESCRIPTION
**Pull request by jiangxt2**
### What changes were proposed in this pull request?

This pull request fixes ClickHouse table-level SETTINGS readback when a scalar literal contains commas.

- Replace the unconditional comma split in `ClickHouseTableOperations.parseSettingsClause()` with top-level delimiter parsing that ignores commas and equals signs inside quoted text and nested parentheses.
- Locate the top-level `SETTINGS` clause while ignoring quoted text and engine parameters, and preserve the complete remainder of the server-returned `engine_full` value.
- Preserve the complete server-returned scalar-literal text and the existing `settings.*` property representation, including the current last-value-wins behavior for duplicate keys.
- Reuse the existing ClickHouse quote and parenthesis scanning behavior for backslash escapes, doubled quote delimiters, and quoted identifiers.
- Reject structurally malformed SETTINGS metadata with a setting-specific error that does not include the raw value.
- Add focused unit coverage and a real ClickHouse native-load and load/recreate regression test.

No public API, property key, dependency, documentation, ALTER behavior, or general SQL parsing capability is changed.

### Why are the changes needed?

The ClickHouse catalog reads table-level settings from `system.tables.engine_full`, but the current parser splits the SETTINGS clause at every comma. A valid String setting such as `merge_workload = 'gravitino,quoted,comma'` is therefore loaded as only `'gravitino`, even though ClickHouse preserves the complete literal. Recreating a table from the loaded property cannot retain the original value.

The existing write path already treats each `settings.*` value as one complete ClickHouse scalar literal, so this change restores read/write symmetry without changing the property contract.

Fix: #12913

### Does this PR introduce _any_ user-facing change?

Yes. Loading a ClickHouse table now preserves complete table-level setting values when quoted text or nested function-style values contain commas. Existing simple settings and `settings.*` property keys are unchanged. Structurally malformed SETTINGS metadata now fails explicitly instead of being silently ignored or converted into truncated properties.

### How was this patch tested?

- `./gradlew :catalogs-contrib:catalog-jdbc-clickhouse:spotlessCheck` — passed on the current revision `e71a4518`.
- `./gradlew :catalogs-contrib:catalog-jdbc-clickhouse:test --tests 'org.apache.gravitino.catalog.clickhouse.operations.TestClickHouseTableOperationsUnit' -PskipITs` — passed on the current revision, covering quoted `COMMENT` values, quoted engine parameters containing `SETTINGS`, malformed metadata before the SETTINGS clause, escaping, doubled quotes, nested parentheses, and duplicate keys.
- `git diff --check` — passed.
- Docker-tagged ClickHouse integration tests were not rerun after the review fixes because the local macOS `docker-connector` was unavailable; no current-revision Docker IT result is claimed.

### CODE REVIEW

File: catalogs-contrib/catalog-jdbc-clickhouse/src/main/java/org/apache/gravitino/catalog/clickhouse/operations/ClickHouseTableOperations.java [hunk 1]
```diff
@@ -1492,25 +1494,53 @@ private ShowCreateTableMetadata parseCreateStatement(String createSql) {
...
-        String key = trimmed.substring(0, eqIdx).trim();
-        String value = trimmed.substring(eqIdx + 1).trim();
-        settings.put(TableConstants.SETTINGS_PREFIX + key, value);
+    int fragmentStart = 0;
+    int equalsIndex = -1;
+    for (int i = 0; i < settingsStr.length(); i++) {
+      char current = settingsStr.charAt(i);
+      if (isQuoteDelimiter(current)) {
+        int quoteEnd = findClosingQuote(settingsStr, i);
+        Preconditions.checkArgument(quoteEnd >= 0, INVALID_SETTINGS_METADATA_MSG);
```
Comment (yuqi1129) [inline catalogs-contrib/catalog-jdbc-clickhouse/src/main/java/org/apache/gravitino/catalog/clickhouse/operations/ClickHouseTableOperations.java:1505]: [P2] Make SETTINGS clause extraction quote-aware before enforcing structural validation.

`parseSettingsFromEngineFull()` still uses `SETTINGS_PATTERN`, which matches `SETTINGS` and terminates at `COMMENT` even inside quoted text. The new validation therefore rejects valid server metadata. For example, ClickHouse 24.8.14 accepts:

```sql
CREATE TABLE quoted_comment (id Int32)
ENGINE = MergeTree ORDER BY id
SETTINGS merge_workload = 'gravitino,COMMENT,comma';
```

Its `engine_full` is:

```text
MergeTree ORDER BY id SETTINGS merge_workload = 'gravitino,COMMENT,comma', index_granularity = 8192
```

The regex passes only `merge_workload = 'gravitino,` to this method, so this check throws `IllegalArgumentException: Invalid ClickHouse table SETTINGS metadata`. That propagates through `getSystemTableMetadata()` and aborts `loadTable()`.

The start boundary has the same issue: a valid `ReplacingMergeTree` table with a UInt64 version column named `SETTINGS version` produces:

```text
ReplacingMergeTree(`SETTINGS version`) ORDER BY id SETTINGS index_granularity = 8192
```

The regex starts inside the quoted identifier, and parsing fails at this same line.

I reproduced both native table definitions with the project's ClickHouse 24.8.14 image, then passed the returned metadata to the actual PR parser in two supplemental regression tests; both fail here. The existing 103 unit tests pass. The regex bug predates this change, but the new strict validation turns incorrect property readback into a table-load failure.

Please locate the clause boundaries with quote/parenthesis-aware scanning as well, and add these cases to the regression coverage.

Comment (jiangxt2) [reply catalogs-contrib/catalog-jdbc-clickhouse/src/main/java/org/apache/gravitino/catalog/clickhouse/operations/ClickHouseTableOperations.java:1505]: Thanks for the detailed review. Both `engine_full` samples from your comment now parse correctly. Addressed in commit `e71a4518`.

`parseSettingsFromEngineFull()` now locates only the top-level `SETTINGS` clause, skipping quoted text and balanced engine parameters. Based on ClickHouse's `ASTStorage` formatting contract, `SETTINGS` is the final storage clause, so the parser consumes the remainder of the string instead of relying on a regex `COMMENT` terminator. This invariant follows ClickHouse's [`ASTStorage::formatImpl`](https://github.com/ClickHouse/ClickHouse/blob/2497bf5356d42e9bd263af6dd4d20e59508b53b8/src/Parsers/ASTCreateQuery.cpp).

Malformed metadata before the `SETTINGS` clause now raises the same fixed validation error instead of silently returning an empty settings map.

The added coverage includes:

- `COMMENT` inside a quoted setting value;
- `SETTINGS` inside quoted engine identifiers and engine parameters;
- unbalanced parentheses;
- unterminated quoted text;
- stray closing parentheses.

Spotless, the focused unit test, and `git diff --check` passed on the current revision. The Docker-tagged ClickHouse IT has not been rerun for this revision because the local macOS `docker-connector` is unavailable, so no integration-test pass is being claimed.

File: catalogs-contrib/catalog-jdbc-clickhouse/src/main/java/org/apache/gravitino/catalog/clickhouse/operations/ClickHouseTableOperations.java [hunk 2]
```diff
@@ -1468,34 +1468,97 @@ Transform[] parsePartitioning(@Nullable String partitionKey) {
...
       return Collections.emptyMap();
     }
 
-    Matcher settingsMatcher = SETTINGS_PATTERN.matcher(engineFull);
-    if (settingsMatcher.find()) {
-      return parseSettingsClause(settingsMatcher.group(1));
+    // engine_full is formatted from ClickHouse's ASTStorage, where SETTINGS is the final storage
+    // clause. Locate it at top level and parse the remainder so keywords in engine parameters and
+    // quoted values are not treated as clause boundaries.
+    int settingsStart = findTopLevelKeyword(engineFull, "SETTINGS");
```
Comment (yuqi1129) [inline catalogs-contrib/catalog-jdbc-clickhouse/src/main/java/org/apache/gravitino/catalog/clickhouse/operations/ClickHouseTableOperations.java]: [P2] Keep a top-level `COMMENT` boundary after `SETTINGS`

The old pattern stopped the clause at `\bCOMMENT\b`; this now hands everything after `SETTINGS` to `parseSettingsClause`. I agree that `engine_full` is formatted from `ASTStorage` and the table comment lives on `ASTCreateQuery`, so today SETTINGS is the last clause. But if any server version does append `COMMENT '...'` here, the result is not an error but a silently corrupted value: `index_granularity = 8192 COMMENT 'x'` parses as one fragment with value `8192 COMMENT 'x'`, and a recreate would send that back to ClickHouse.

A cheap guard keeps the old behaviour without the old regex's weakness: run `findTopLevelKeyword(remainder, "COMMENT")` on the text after `SETTINGS` and cut there when found. Quoted `COMMENT` inside a value is already skipped by the top-level scan, so `'id,COMMENT,name,val'` in the new test stays intact.

Related nit: `findTopLevelKeyword` returns the first top-level match. A column literally named `settings` is emitted unquoted by `backQuoteIfNeed`, so `ORDER BY settings SETTINGS index_granularity = 8192` starts parsing at the column and yields the key `SETTINGS index_granularity`. The old regex had the same edge, but since SETTINGS is the final storage clause, taking the last top-level match fixes it for free.

The quote scanning itself (`findClosingQuote` with backslash and doubled-delimiter escapes for all three quote characters) matches the ClickHouse lexer and what `quoteString`/`backQuoteIfNeed` emit; I checked the edge cases (trailing lone backslash, `'a\\'`, `'a''` unterminated) and they behave correctly.

**Review by yuqi1129 — COMMENTED** [review]
One suggestion inline; the quote handling looks correct to me.

File: catalogs-contrib/catalog-jdbc-clickhouse/src/main/java/org/apache/gravitino/catalog/clickhouse/operations/ClickHouseTableOperations.java (hunk 2 above)
Comment (jiangxt2) [reply catalogs-contrib/catalog-jdbc-clickhouse/src/main/java/org/apache/gravitino/catalog/clickhouse/operations/ClickHouseTableOperations.java]: Addressed in commit aa5dc48f: the parser now selects the last top-level SETTINGS, with unit and native-load regression coverage.

I have not reproduced a top-level `COMMENT` suffix in `engine_full` on ClickHouse 24.8.14.39; table comments appear in `create_table_query` instead. I therefore left this guard out of the current PR. If a  supported version reproduces it, I can address it in a follow-up PR.

**Review by yuqi1129 — APPROVED** [review]
